### PR TITLE
Semperis directory services protector

### PR DESCRIPTION
**Pull request by ericlanteigne-semperis**
## Proposed Changes

  - Adding Semperis Directory Services Protector connector

### CODE REVIEW

**Comment by ericlanteigne-semperis** [comment]
We are using the Event table populated by an agent collecting windows event logs. It seems there is no schema available in the tests for the Event table. The DataConnectorValidations are not listed as required but I would be happy to update the tests accordingly. If you have advice on how to get the schema for the Event table that would be very useful.

File: Solutions/SemperisDirectoryServicesProtector/Workbooks/Images/Logo/Semperis.svg [hunk 1]
```diff
@@ -0,0 +1 @@
...V76a1.92,1.92,0,0,1-1.3,1.8L37.7,86a2.2,2.2,0,0,1-.8.1A2.77,2.77,0,0,1,36,86L12.5,77.8A1.92,1.92,0,0,1,11.2,76l.1-8.5L.1,65.6V82.2a3.22,3.22,0,0,0,2.1,3L35.9,98a3.55,3.55,0,0,0,1,.2,2.92,2.92,0,0,0,.9-.2L71.5,85.2a3.22,3.22,0,0,0,2.1-3V61.8A3.51,3.51,0,0,0,70.8,58.6Z" style="fill:url(#a)"/><path d="M2.7,39.7l34.1,5.7,36.7,6.1V43a3.22,3.22,0,0,0-2.7-3.2L36.8,34,12.9,30a2,2,0,0,1-1.6-1.9V22.3a1.92,1.92,0,0,1,1.3-1.8l23.3-8.2a4.1,4.1,0,0,1,1.8,0l23.5,8.2a1.92,1.92,0,0,1,1.3,1.8l-.1,8.5,11.2,1.9V16a3.22,3.22,0,0,0-2.1-3L37.6.2a2.12,2.12,0,0,0-1.8,0L2.1,13A3.22,3.22,0,0,0,0,16V36.5A3.22,3.22,0,0,0,2.7,39.7Z" style="fill:url(#b)"/><g style="opacity:0.15;mix-blend-mode:multiply"><path d="M70.8,58.6,36.85,52.91V64.29h0l23.9,4a2,2,0,0,1,1.6,1.9V76a1.92,1.92,0,0,1-1.3,1.8L37.7,86a2.2,2.2,0,0,1-.8.1h0V98.19h0a2.92,2.92,0,0,0,.9-.2L71.5,85.2a3.22,3.22,0,0,0,2.1-3V61.8A3.51,3.51,0,0,0,70.8,58.6Z" style="fill:url(#c)"/><path d="M73.5,43a3.22,3.22,0,0,0-2.7-3.2L36.85,34v11.4L73.5,51.5Z" style="fill:url(#d)"/><path d="M71.5,13,37.6.2A4.09,4.09,0,0,0,36.85,0V12.2a2.75,2.75,0,0,1,.85.1l23.5,8.2a1.92,1.92,0,0,1,1.3,1.8l-.1,8.5,11.2,1.9V16A3.22,3.22,0,0,0,71.5,13Z" style="fill:url(#e)"/></g></g></svg>
```
Comment (v-admahe) [inline Solutions/SemperisDirectoryServicesProtector/Workbooks/Images/Logo/Semperis.svg:1]: Id in logo file should be of guid type and xmlns:xlink should not be there in the file.

Please follow the guideline for logo [here](https://github.com/Azure/Azure-Sentinel/blob/e1665d4c4442b430d6847adc65f4b9c25f7b071f/DataConnectors/ReadMe.md)

Thanks,
Aditi Maheshwari

**Review by v-admahe — CHANGES_REQUESTED** [review]

File: Solutions/SemperisDirectoryServicesProtector/Workbooks/Images/Logo/Semperis.svg (hunk 1 above)
Comment (v-admahe) [reply Solutions/SemperisDirectoryServicesProtector/Workbooks/Images/Logo/Semperis.svg:1]: @ericlanteigne-semperis Could ypu please add the sample data in the PR in json format.

Thanks,
Aditi Maheshwari

**Comment by oshezaf** [comment]
Few comments:

- The parser uses mv-expand breaking each event to multiple events with a single name value pair from the XML. It is not a valid parser, even if it may work for the simplistic detections currently available. Also, parse_xml is highly inefficient. The parser should extract the relevant fields needed, preferably using parse.
- The detections use frequency of 1h and period of 5h. This will result in duplication.  the frequency and period should be the same.
- Since all detections are identical, essentially elevating an event to an alert, they should use a single rule and the new dynamic alert capabitlites.
- Is the summarization in detections intentional? Trying to solve the parser issues? If the intent is to raise an alert for each event, as I think it should be, Event grouping should be used.
- The connector lists WindowsEvent as data type while it should be Event.

File: Solutions/SemperisDirectoryServicesProtector/Workbooks/Images/Logo/Semperis.svg (hunk 1 above)
Comment (v-admahe) [reply Solutions/SemperisDirectoryServicesProtector/Workbooks/Images/Logo/Semperis.svg:1]: @ericlanteigne-semperis Please move the workbook logo file in and preview images in solution folder you can take reference of folder structure from [here](https://github.com/Azure/Azure-Sentinel/tree/master/Solutions/PaloAltoPrismaCloud/Workbooks/Images). and also if its a solution workbook metadata is not needed, so could you please revert the changes of workbook metadata.

Thanks,
Aditi Maheshwari

**Comment by v-admahe** [comment]
@ericlanteigne-semperis Please move the workbook logo file in and preview images in solution folder you can take reference of folder structure from here. and also if its a solution workbook metadata is not needed, so could you please revert the changes of workbook metadata.

Thanks,
Aditi Maheshwari

**Comment by v-admahe** [comment]
@ericlanteigne-semperis Could ypu please add the sample data in the PR in json format.

Thanks,
Aditi Maheshwari

**Comment by ericlanteigne-semperis** [comment]
I will reopen this pull request when we are done recommended changes

File: Solutions/SemperisDirectoryServicesProtector/Analytic Rules/SemperisDSP_EvidenceOfMimikatzDCShadowAttack.yaml [hunk 2]
```diff
@@ -0,0 +1,23 @@
+id: 1a6d0a49-64b3-4ca1-96c3-f154c16c218c
+name: Semperis DSP Mimikatz's DCShadow Alert
+description: |
+  'Mimikatz's DCShadow switch allows a user who has compromised an AD domain, to inject arbitrary changes into AD using a "fake" domain controller. These changes bypass the security event log and can't be spotted using normal AD tools. This rule looks for evidence that a machine has been used in this capacity.'
+severity: High
+requiredDataConnectors:
+  - connectorId: SemperisDSP
+    dataTypes:
+      - WindowsEvent
```
[the remark this answers is not in the document]
Comment (ericlanteigne-semperis) [reply Solutions/SemperisDirectoryServicesProtector/Analytic Rules/SemperisDSP_EvidenceOfMimikatzDCShadowAttack.yaml]: Can we resolve this comment?

File: Solutions/SemperisDirectoryServicesProtector/Analytic Rules/SemperisDSP_KerberoskrbtgtAccount.yaml [hunk 3]
```diff
@@ -0,0 +1,23 @@
+id: 9ff3b26b-7636-412e-ac46-072b084b94cb
+name: Semperis DSP Kerberos krbtgt account with old password
+description: |
+  'The krbtgt user account is a special (disabled) user account in every Active Directory domain that has a special role in Kerberos function. If this account's password is compromised, Golden Ticket attacks can be performed to get access to any resource in the AD domain. This indicator looks for a krbtgt user account whose password hasn't been changed in the past 180 days. While Microsoft recommends changing the password every year, STIG recommends changing it every 180 days.'
+severity: Medium
+requiredDataConnectors:
+  - connectorId: SemperisDSP
+    dataTypes:
+      - WindowsEvent
```
[the remark this answers is not in the document]
Comment (ericlanteigne-semperis) [reply Solutions/SemperisDirectoryServicesProtector/Analytic Rules/SemperisDSP_KerberoskrbtgtAccount.yaml]: Can we resolve this comment?

File: Solutions/SemperisDirectoryServicesProtector/Analytic Rules/SemperisDSP_RecentsIDHistoryChangesOnADObjects.yaml [hunk 4]
```diff
@@ -0,0 +1,23 @@
+id: 64796da3-6383-4de2-9c97-866c83c459ae
+name: Semperis DSP Recent sIDHistory changes on AD objects
+description: |
+  'This indicator detects any recent changes to sIDHistory on AD objects, including changes to non-privileged accounts where privileged SIDs are added.'
+severity: High
+requiredDataConnectors:
+  - connectorId: SemperisDSP
+    dataTypes:
+      - WindowsEvent
```
[the remark this answers is not in the document]
Comment (ericlanteigne-semperis) [reply Solutions/SemperisDirectoryServicesProtector/Analytic Rules/SemperisDSP_RecentsIDHistoryChangesOnADObjects.yaml]: Can we resolve this comment?

File: Solutions/SemperisDirectoryServicesProtector/Analytic Rules/SemperisDSP_WellKnownPrivilegedSIDsInsIDHistory.yaml [hunk 5]
```diff
@@ -0,0 +1,24 @@
+id: ddd75d93-5b8b-4349-babe-c4e15343c5a3
+name: Semperis DSP Well-known privileged SIDs in sIDHistory
+description: |
+  'This indicator looks for security principals that contain specific SIDs of accounts from built-in privileged groups within their sIDHistory attribute. This would allow those security principals to have the same privileges as those privileged accounts, but in a way that is not obvious to monitor (e.g. through group membership).'
+severity: Medium
+requiredDataConnectors:
+  - connectorId: SemperisDSP
+    dataTypes:
+      - WindowsEvent
```
[the remark this answers is not in the document]
Comment (ericlanteigne-semperis) [reply Solutions/SemperisDirectoryServicesProtector/Analytic Rules/SemperisDSP_WellKnownPrivilegedSIDsInsIDHistory.yaml]: Can we resolve this comment?

File: Solutions/SemperisDirectoryServicesProtector/Analytic Rules/SemperisDSP_ZerologonVulnerability.yaml [hunk 6]
```diff
@@ -0,0 +1,23 @@
+id: 85c1f9e4-6f14-46bf-82d5-dbe495b92aab
+name: Semperis DSP Zerologon vulnerability
+description: |
+  'This indicator looks for security vulnerability to CVE-2020-1472, which was patched by Microsoft in August 2020. Without this patch, an unauthenticated attacker can exploit CVE-2020-1472 to elevate their privileges and get administrative access on the domain.'
+severity: Medium
+requiredDataConnectors:
+  - connectorId: SemperisDSP
+    dataTypes:
+      - WindowsEvent
```
[the remark this answers is not in the document]
Comment (ericlanteigne-semperis) [reply Solutions/SemperisDirectoryServicesProtector/Analytic Rules/SemperisDSP_ZerologonVulnerability.yaml]: Can we resolve this comment?

File: Solutions/SemperisDirectoryServicesProtector/Data Connectors/SemperisDSP-connector.json [hunk 7]
```diff
@@ -0,0 +1,152 @@
+{
+    "id": "SemperisDSP",
+    "title": "Semperis Directory Services Protector",
+    "publisher": "SEMPERIS",
+    "descriptionMarkdown": "Semperis Directory Services Protector data connector allows for the export of its Windows event logs (i.e. Indicators of Exposure and Indicators of Compromise) to Azure Sentinel in real time.\nIt provides a data parser to manipulate the Windows event logs more easily. The different workbooks ease your Active Directory seurity monitoring and provide different ways to visualize the data. The analytic templates allow to automate responses regarding different events, exposures, or attacks.",
+    "additionalRequirementBanner": "This data connector depends on a parser based on Kusto Function to work as expected. Follow the steps to use this Kusto Function alias **dsp_parser** in queries and workbooks. [Follow steps to get this Kusto Function>](https://github.com/Azure/Azure-Sentinel/tree/master/Solutions/SemperisDirectorySerivesProtector/Parsers/dsp_parser.txt)  ",
```
[the remark this answers is not in the document]
Comment (ericlanteigne-semperis) [reply Solutions/SemperisDirectoryServicesProtector/Data Connectors/SemperisDSP-connector.json]: Can we resolve this comment?

File: Solutions/SemperisDirectoryServicesProtector/Data Connectors/SemperisDSP-connector.json [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+                "requiredPermissions": {
+                    "action": true
+                }
+            }
+        ]
+    },
+    "instructionSteps": [
+        {
+            "title": "", 
+            "description": ">This data connector depends on a parser based on a Kusto Function to work as expected. [Follow these steps](https://github.com/Azure/Azure-Sentinel/tree/master/Solutions/SemperisDirectorySerivesProtector/Parsers/dsp_parser.txt) to create the Kusto Functions alias, **dsp_parser**",
```
[the remark this answers is not in the document]
Comment (ericlanteigne-semperis) [reply Solutions/SemperisDirectoryServicesProtector/Data Connectors/SemperisDSP-connector.json]: Can we resolve this comment?

**Comment by ericlanteigne-semperis** [comment]
@v-admahe I've added the Sample Data. I can see that most of the other sample data files are CSV. I added that one as well since I wasn't sure. In the ReadMe for Solutions it says to put it in the appropriate folder but seeing as our parser is using the Event table, I am not certain where it should go. Do you have advice?
Also I removed the changes from Workbooks directory and kept everything in the Solutions directory.

**Comment by ericlanteigne-semperis** [comment]
@oshezaf Do you have more information on "the new dynamic alert capabilities"? I wasn't quite able to find information on that just yet. That would be helpful, thanks. 
At the moment we just want a single alert to be triggered when the detection query has results, so I've removed the summarize statement. Based on your message it sounds like there is a way to trigger multiple alerts when running a single detection, is there more information about this in documentation? I am not seeing information stating that a single rule can have multiple alerts but I am curious about it. 
I am reviewing how the parser is working to see if I can use parse and avoid mx-expand however the information in EventData does not have a defined schema so I am not certain we will be able to make that change to parse_xml. Is there an issue if we have to use parse_xml? 
Also, we are using pivot to recombine the results that were expanded via mx-expand based on the values that were common before the expansion. Is this not an adequate method of maintaining data consistency?

**Comment by oshezaf** [comment]
@ericlanteigne-semperis :

- Event grouping (i.e. generating an alert per query result): Just added documentation on that here: https://github.com/Azure/Azure-Sentinel/wiki/Query-Style-Guide#eventgroupingsettings.

- Pivot will not actually regenerate the events. It will group them based on the dimensions you specify, but since there is no unique key dimension, it is not deterministic. In practice you pivot on time with the risk of two events at the same time being grouped together. It is also highly inefficient for parsers and will break for high rates of events. I am aware there are other parsers using this method, but it is flawed. As to the alternative: parse is preferable for any part of the XML that has repeating structure. You can have multiple parse statements and use extract/extract_all for variable parts. Can you share XML samples?

**Comment by ericlanteigne-semperis** [comment]
@oshezaf Thank you for the information. 

I included sample data in both csv and json format. I've simply put it in the Sample Data directory for now. @v-admahe asked for it but I still need to confirm where exactly I should put it

**Comment by ericlanteigne-semperis** [comment]
I have an issue open with regards to the WorkbooksValidations failing.
https://github.com/Azure/Azure-Sentinel/issues/2389

Also, the sample data I added is causing the DocumentsLinkValidation to fail. Any advice on where that should be located? Inside the Solution folder potentially?

Lastly I haven't gotten any feedback regarding the DataConnectorValidations that are failing. I included a comment originally with regards to this, if you have any advice on this as well that would be much appreciated

> We are using the Event table populated by an agent collecting windows event logs. It seems there is no schema available in the tests for the Event table. The DataConnectorValidations are not listed as required but I would be happy to update the tests accordingly. If you have advice on how to get the schema for the Event table that would be very useful.

Thank you

File: Solutions/SemperisDirectoryServicesProtector/Parsers/dsp_parser.txt [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+// REFERENCE:
+// Using functions in Azure monitor log queries: https://docs.microsoft.com/azure/azure-monitor/log-query/functions
+//
+//
+Event
+| where Source == "Semperis-DSP-Security"
+| extend EventData = parse_xml(EventData).DataItem.EventData.Data
+| mv-expand bagexpansion=array EventData
+| evaluate bag_unpack(EventData)
+| extend Key=tostring(['@Name']), Value=['#text']
```
[the remark this answers is not in the document]
Comment (ericlanteigne-semperis) [reply Solutions/SemperisDirectoryServicesProtector/Parsers/dsp_parser.txt]: The parser has been updated and this is no longer used. Can this comment be resolved?

Thanks

File: Solutions/SemperisDirectoryServicesProtector/Workbooks/Images/Logo/Semperis.svg (hunk 1 above)
Comment (ericlanteigne-semperis) [reply Solutions/SemperisDirectoryServicesProtector/Workbooks/Images/Logo/Semperis.svg]: I have done these changes. Please let me know if I have missed anything.

Thank you

**Comment by ericlanteigne-semperis** [comment]
Going to open a new clean pull request since this one seems to have gotten too old and nobody is replying to my comments

**Pull request closed without merging**